lambdas/bulk-migrate-vcs/src/main/java/uk/gov/di/ipv/core/bulkmigratevcs/BulkMigrateVcsHandler.java

```diff
@@ -0,0 +1,271 @@
+                                .with(LOG_BATCH_ID.getFieldName(), batchSummary.getBatchId()));
+                // Send migrated audit event
+                batchSummary.incrementMigrated();
+            } catch (Throwable e) {
+                logError(
+                        "Migration failed - error writing to tactical",
+                        reportItem,
+                        e,
+                        batchSummary.getBatchId());
+                // Send failed audit event
+                batchSummary.incrementFailedTacticalWrite(reportItem.getHashUserId());
+            }
+        } else {
+            LOGGER.info(
+                    LogHelper.buildLogMessage(
+                                    String.format("Skipping migration - %s", getSkippedReason(vcs)))
+                            .with(LOG_HASH_USER_ID.getFieldName(), reportItem.getHashUserId())
+                            .with(LOG_BATCH_ID.getFieldName(), batchSummary.getBatchId()));
+            // Send skipped audit event
+            batchSummary.incrementSkipped();
+        }
+    }
+
+    private String getSkippedReason(List<VerifiableCredential> vcs) {
+        return vcs.isEmpty() ? "no VCs to migrate" : "all VCs already migrated";
+    }
+
+    private String getBatchId(String lastEvaluatedKey, int pageSize) {
+        // The last evaluated key with the number of items in the page - should be enough to
+        // process the same batch in future
+        return String.format(
+                "%s:%d", lastEvaluatedKey == null ? "START" : lastEvaluatedKey, pageSize);
+    }
+
+    private void storeUnmigratedVcsInEvcs(
+            String userId, List<VerifiableCredential> vcs, String batchId, Instant timestamp)
+            throws EvcsServiceException {
+        evcsClient.storeUserVCs(
+                userId,
+                vcs.stream()
+                        .filter(vc -> vc.getMigrated() == null)
+                        .map(
+                                vc ->
+                                        new EvcsCreateUserVCsDto(
+                                                vc.getVcString(),
+                                                CURRENT,
+                                                new EvcsMetadata(batchId, timestamp.toString()),
+                                                MIGRATED))
+                        .toList());
+    }
+
+    private void setMigratedOnTacticalStoreVcs(List<VerifiableCredential> vcs, Instant timestamp)
+            throws VerifiableCredentialException {
+        verifiableCredentialService.updateIdentity(
+                vcs.stream()
+                        .filter(vc -> vc.getMigrated() == null)
+                        .map(
+                                vc -> {
+                                    vc.setMigrated(timestamp);
+                                    return vc;
+                                })
```

Comment:
That was my original implementation, but then sonar shouted at me for using a method that's supposed to be only meant for debugging. Apparently it can be skipped by the JVM for optimisation reasons. Although if that'd actually happen when you've got it doing "stuff" I'm not sure.

Reply:
It sounds like the Sonar rule is slightly misguided, the point is that streams are lazy, and intermediate operations will only occur when they are needed (rather than them being dropped arbitrarily). For example:
```
Stream.of("foo", "bar", "baz")
    .peek((v) -> System.out.println(v))
    .findFirst();
```
This will (probably) only print the first entry, but the same would apply if it was done in a `.map` as well! (They are both intermediate operations).

The additional warning applies to `peek`, because you're more likely to do something side-effecty. (With a `map`, this behaviour is probably a bit more expected)

Reply:
In this case, the optimal implementation is probably to use neither of these options:
```
vcs.forEach(vc -> vc.setMigrated(timestamp));
verifiableCredentialService.updateIdentity(vcs);
```
😄